src/main/java/ecdar/backend/BackendDriver.java

```diff
@@ -65,12 +65,16 @@ private BackendConnection getBackendConnection(BackendInstance backend) throws B
                 tryStartNewBackendConnection(backend);
             }
 
-            // Block until a connection becomes available
-            connection = openBackendConnections.get(backend).take();
+            if (backend.isThreadSafe()){
+                connection = openBackendConnections.get(backend).peek();
```

Comment:
The queue seems pretty useless when using a thread safe engine. Once the thread safe engine is in the top of the queue, it will always be used and no other engine instances will be used?